docs/test-suite/writing-portable-tests.md

```diff
@@ -0,0 +1,541 @@
+
+    # Now, we check if the pattern 'Sum of all ranks: X' with X the correct sum for the amount of ranks is found
+    # in the standard output:
+    # https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#reframe.core.builtins.sanity_function
+    @sanity_function
+    def validate(self):
+        # Sum of 0, ..., N-1 is (N * (N-1) / 2)
+        sum_of_ranks = round(self.scale * ((self.scale - 1) / 2))
+        # https://reframe-hpc.readthedocs.io/en/stable/deferrable_functions_reference.html#reframe.utility.sanity.assert_found
+        return sn.assert_found(r'Sum of all ranks: %s' % sum_of_ranks, self.stdout)
+
+    # Now, we define a pattern to extract a number that reflects the performance of this test
+    # https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#reframe.core.builtins.performance_function
+    @performance_function('s')
+    def time(self):
+        # https://reframe-hpc.readthedocs.io/en/stable/deferrable_functions_reference.html#reframe.utility.sanity.extractsingle
+        return sn.extractsingle(r'^Time elapsed:\s+(?P<perf>\S+)', self.stdout, 'perf', float)
+```
+
+This single test class will generate 6 test instances: tests with 2, 128 and 256 tasks for each of the two modules, respectively. It will check the sum of ranks produced at the end in the output, which is how ReFrame will validate that the test ran correctly. Finally, it will also print the performance number that was extracted by the `performance_function`.
+
+This test _works_, but is _not_ very portable. If we move to a system with 192 cores per node, the current `scale` parameter is a bit awkward. The test would still run, but we wouldn't have a test instance that just tests this on a full (single) node or two full nodes. Furthermore, if we add a new `mpi4py` module in EESSI, we would have to alter the test to add the name to the list, since the module names are hardcoded in this test.
+
+### Step 3: implementing as a portable ReFrame test { #as-portable-reframe-test }
+
+In the previous section, there were several system-specific items in the test. In this section, we will show how we use the EESSI hooks to avoid hard-coding system specific information. We do this by replacing the system-specific parts of the test from Step 2 bit by bit. The full final test can be found under `tutorials/mpi4py/mpi4py_portable.py` in the [EESSI test suite](https://github.com/EESSI/test-suite/) repository.
+
+#### Replacing hard-coded test scales (mandatory)
+
+We replace the hard-coded
+
+```python
+    # ReFrame will generate a test for each scale
+    scale = parameter([2, 128, 256])
+```
+
+by 
+
+```python
+from eessi.testsuite.constants import SCALES
+...
+    # ReFrame will generate a test for each scale
+    scale = parameter(SCALES.keys())
+```
+
+the `SCALES` constant (TODO: API reference) contains a set of default scales at which we run all tests. For our `mpi4py` example, that is sufficient. 
```

Comment:
@casparvl Maybe remove the `TODO` for now, looks a bit awkward in the rendered docs, and we don't want to block this PR by not having API docs yet
```suggestion

the `SCALES` constant contains a set of default scales at which we run all tests. For our `mpi4py` example, that is sufficient.

Reply:
True, they were there for my own sake. Can we leave a comment that doesn't get rendered? That would be helpful to our future selves...

Reply:
You could use HTML-style comments, like:

```
<!-- TODO add API reference -->
```

but I'm not sure it's worth it. Once we have the API docs, we'll need to go through this whole page again, I think, since the TODO didn't seem consistent to me (not all hooks, etc.)

Reply:
Fair enough, I now replaced everything with references to code :)